File: generate-pdf-worker/utils/consumer_circuitbreaker.py
import time
import traceback
from .logger import log
logger = log(service="generate-pdf")

class ConsumerCircuitBreaker:
    def __init__(self):
        self.failureCount = 0
        self.state = "CLOSED"
        self.threshold = 5
        self.lastFailureTime = 0
        self.timeout = 60

    def execute(self, toExecute):
        if self.state == "OPEN":
            if time.time() - self.lastFailureTime > self.timeout:
                self.state = "HALF_OPEN"
                logger.info("Circuit state has been set to HALF_OPEN")
            else:
                logger.info("Circuit OPEN - Can't process requests currently")
                raise
        
        try:
            result = toExecute()
            self.onSuccess()
            return result
        except Exception as e:
            tb = traceback.format_exc()
            logger.error("function execution failed \n {error} \n {traceback}", error=str(e), traceback=tb)
            self.onFailure()
            raise
            

    def onSuccess(self):
        self.failureCount = 0
        self.state = "CLOSED"
        pass

    def onFailure(self):
        self.failureCount+=1
        self.lastFailureTime = time.time()

        if self.failureCount >= self.threshold:
            self.state = "OPEN"


    def getState(self):
        return { 
                "state": self.state, 
                "failureCount": self.failureCount, 
                "threshold": self.threshold,
                "timeSinceLastFailure": time.time() - self.lastFailureTime if self.lastFailureTime > 0 else 0
                }
    
    def getTimeSinceLastFailure(self):
        return time.time() - self.lastFailureTime if self.lastFailureTime > 0 else 0
    
circuitbreaker = ConsumerCircuitBreaker()

File: resize-image/utils/consumer_circuit_breaker.go
package utils

import (
	"errors"
	"sync"
	"time"
)

type CircuitBreaker struct {
	mu           sync.RWMutex
	failureCount int
	state        string
	threshold    int
	timeout      time.Duration
	lastFailTime time.Time
}

func NewCircuitBreaker(threshold int, timeout time.Duration) *CircuitBreaker {
	return &CircuitBreaker{
		state:     "CLOSED",
		threshold: threshold,
		timeout:   timeout,
	}
}

func (cb *CircuitBreaker) Execute(fn func() error) error {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if cb.state == "OPEN" {
		if time.Since(cb.lastFailTime) > cb.timeout {
			cb.state = "HALF_OPEN"
		} else {
			return errors.New("circuit breaker is OPEN")
		}
	}

	err := fn()
	if err != nil {
		cb.onFailure()
		return err
	}

	cb.onSuccess()
	return nil
}

func (cb *CircuitBreaker) onSuccess() {
	cb.failureCount = 0
	cb.state = "CLOSED"
}

func (cb *CircuitBreaker) onFailure() {
	cb.failureCount++
	cb.lastFailTime = time.Now()

	if cb.failureCount >= cb.threshold {
		cb.state = "OPEN"
	}
}
